src/carousel/index.ts

```diff
@@ -0,0 +1,202 @@
+    const currentSlide = from < to ? to % length : (to % length + length) % length;
+    this.set({ currentSlide });
+  }
+
+  disableTransition = () => {
+    this.refs.track.removeEventListener('transitionend', this.disableTransition, false);
+    this.set({ transitioning: false });
+  }
+
+  trackStyle = () => {
+    const { speed } = this.props;
+    if (this.state.transitioning && speed) {
+      const transition = `${speed}ms ease`;
+      return {
+        ...this.getStaticTrackStyle(),
+        transition,
+        '-webkit-transition': transition,
+        '-ms-transition': transition
+      };
+    } else {
+      return this.getStaticTrackStyle();
+    }
+  }
+
+  getStaticTrackStyle = () => {
+    const slideWidth = this.getSlideWidth();
+    const { items: { length: slideCount }, slidesToShow } = this.props;
+    // based on total count of products after cloning
+    const trackWidth = (slideCount + 4 * slidesToShow - 2) * slideWidth;
+    const offset = Carousel.calculatePosition(this.state.currentSlide, slideWidth, slidesToShow);
+    const transform = `translateX(-${offset}px)`;
+
+    return {
+      width: `${trackWidth}px`,
+      transform,
+      '-webkit-transform': transform,
+      '-ms-transform': transform
+    };
+  }
+
+  getSlideWidth = () => this.refs.wrapper.offsetWidth / this.props.slidesToShow;
+
+  onTouchStart = (e: TouchEvent & Carousel.Event) => {
+    this.set({ touchObject: { startX: e.touches[0].pageX, startY: e.touches[0].pageY } });
+
+    e.target.addEventListener('touchend', this.onTouchEnd);
+  }
+
+  onTouchEnd = (e: TouchEvent & Carousel.Event) => {
+    const curX = e.changedTouches[0].pageX;
+    const curY = e.changedTouches[0].pageY;
+    const swipeNext = Carousel.shouldSwipeToNext({ ...this.state.touchObject }, curX, curY);
+
+    // swipe distance needs to be more than 20
+    if (Math.abs(curX - this.state.touchObject.startX) > MIN_SWIPE_DISTANCE && typeof swipeNext === 'boolean') {
+      swipeNext ? this.moveNext() : this.movePrevious();
+    }
+
+    e.target.removeEventListener('touchstart', this.onTouchStart);
+    e.target.removeEventListener('touchend', this.onTouchEnd);
+  }
+
+  static calculatePosition = (currentSlide: number, moveDistance: number, slidesToShow: number): number =>
+    // when it first loads, offset the length of cloned items at the begining
+    (currentSlide + slidesToShow * 2 - 1) * moveDistance
+
+  static shouldSwipeToNext = ({ startX, startY }: Carousel.TouchObject, curX: number, curY: number): boolean => {
+    let swipeAngle = Math.round(Math.atan2(startY - curY, startX - curX) * 180 / Math.PI);
+
+    if (swipeAngle < 0) {
+      swipeAngle = 360 + swipeAngle;
+    }
+
+    // swipeAngle between 45 and 135, between 225 and 315 is ignored;
+    // tslint:disable-next-line:max-line-length
+    if ((swipeAngle <= MOVE_NEXT_UPWARD_MAX_ANGLE && swipeAngle >= 0) || (swipeAngle <= 360 && swipeAngle >= MOVE_NEXT_DOWNWARD_MAX_ANGLE)) {
```

Comment:
\>= 0 check is unnecessary,

Is the <=360 check necessary?

Reply:
it is not needed but only for readability.